src/main/java/org/alfresco/repo/security/permissions/impl/SimplePermissionReference.java

```diff
@@ -1,98 +1,99 @@
+ * %%
+ * This file is part of the Alfresco software. 
+ * If the software was purchased under a paid Alfresco license, the terms of 
+ * the paid license agreement will prevail.  Otherwise, the software is 
+ * provided under the following open source license terms:
+ * 
+ * Alfresco is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ * 
+ * Alfresco is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ * 
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Alfresco. If not, see <http://www.gnu.org/licenses/>.
+ * #L%
+ */
 package org.alfresco.repo.security.permissions.impl;
 
-import java.util.HashMap;
-import java.util.concurrent.locks.ReadWriteLock;
-import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Future;
+import java.util.concurrent.FutureTask;
 
+import org.alfresco.error.AlfrescoRuntimeException;
 import org.alfresco.service.namespace.QName;
+import org.alfresco.util.Pair;
 
 /**
  * A simple permission reference.
  * 
  * @author andyh
  */
-public final class SimplePermissionReference extends AbstractPermissionReference
-{   
+public final class SimplePermissionReference extends AbstractPermissionReference {
     private static final long serialVersionUID = 637302438293417818L;
 
-    private static ReadWriteLock lock = new ReentrantReadWriteLock();
-
-    private static HashMap<QName, HashMap<String, SimplePermissionReference>> instances = new HashMap<QName, HashMap<String, SimplePermissionReference>>();
+    //Use thread-safe map initiallized with a slightly larger capacity to reduce the posibility of two or more threads attempting to resize at the same time
+    private static ConcurrentMap<Pair<QName, String>, Future<SimplePermissionReference>> instances = new ConcurrentHashMap<>(100);
 
     /**
-     * Factory method to create simple permission refrences
+     * Factory method to create simple permission references
      * 
      * @return a simple permission reference
      */
-    public static SimplePermissionReference getPermissionReference(QName qName, String name)
-    {
-        lock.readLock().lock();
-        try
+    public static SimplePermissionReference getPermissionReference(QName qName, String name) {
+        //loop if thread is interrupted and should be run again
+        while (true)

```

Comment:
Just wondering if the method could be simplified like this.

` return instances.computeIfAbsent(key, pair -> new SimplePermissionReference(pair.getFirst(), pair.getSecond()));`

Reply:
This actually ends up being slightly slower though I do agree that it is simpler

Reply:
I don't think it is going to be any slower since the computation inside the lambda isn't complex. Also `computeIfAbsent` would avoid unnecessary creation of SimplePermissionReference objects when multiple threads are acting on the same key.